drivers/gpio/gpio_ch32v00x.c

```diff
@@ -0,0 +1,122 @@
+{
+	const struct gpio_ch32v00x_config *config = dev->config;
+	*value = config->regs->INDR;
+	return 0;
+}
+
+static int gpio_ch32v00x_port_set_masked_raw(const struct device *dev, uint32_t mask,
+					     uint32_t value)
+{
+	const struct gpio_ch32v00x_config *config = dev->config;
+	config->regs->BSHR = ((~value & mask) << 16) | (value & mask);
+	return 0;
+}
+
+static int gpio_ch32v00x_port_set_bits_raw(const struct device *dev, uint32_t pins)
+{
+	const struct gpio_ch32v00x_config *config = dev->config;
+	config->regs->BSHR = pins;
+	return 0;
+}
+
+static int gpio_ch32v00x_port_clear_bits_raw(const struct device *dev, uint32_t pins)
+{
+	const struct gpio_ch32v00x_config *config = dev->config;
+	config->regs->BCR = pins;
+	return 0;
+}
+
+static int gpio_ch32v00x_port_toggle_bits(const struct device *dev, uint32_t pins)
+{
+	const struct gpio_ch32v00x_config *config = dev->config;
+	uint32_t changed = (config->regs->OUTDR ^ pins) & pins;
+	config->regs->BSHR = (changed & pins) | (~changed & pins) << 16;
+
+	return 0;
+}
+
+static const struct gpio_driver_api gpio_ch32v00x_driver_api = {
+	.pin_configure = gpio_ch32v00x_configure,
+	.port_get_raw = gpio_ch32v00x_port_get_raw,
+	.port_set_masked_raw = gpio_ch32v00x_port_set_masked_raw,
+	.port_set_bits_raw = gpio_ch32v00x_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_ch32v00x_port_clear_bits_raw,
+	.port_toggle_bits = gpio_ch32v00x_port_toggle_bits,
+};
+
+static int gpio_ch32v00x_init(const struct device *dev)
+{
+	const struct gpio_ch32v00x_config *config = dev->config;
+	clock_control_on(config->clock_dev, (clock_control_subsys_t *)(uintptr_t)config->clock_id);
+
+	return 0;
+}
+
+#define GPIO_CH32V00X_INIT(idx)                                                                    \
+	static const struct gpio_ch32v00x_config gpio_ch32v00x_##idx##_config = {                  \
+		.common = {                                                                        \
+				.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_INST(idx),             \
+			},                                                                         \
+		.regs = (GPIO_TypeDef *)DT_INST_REG_ADDR(idx),                                     \
+		.clock_dev = DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(idx)),                              \
+		.clock_id = DT_INST_CLOCKS_CELL(idx, id),                                          \
+	};                                                                                         \
+                                                                                                   \
+	static struct gpio_ch32v00x_data gpio_ch32v00x_##idx##_data;                               \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(idx, gpio_ch32v00x_init, NULL, &gpio_ch32v00x_##idx##_data,          \
+			      &gpio_ch32v00x_##idx##_config, POST_KERNEL,                          \
```

Comment:
Init of the GPIO's should preferably happen in pre-kernel. I've been working on a UART driver for a ch32V20X device and the GPIO's should be initialized before UART. UART will init and set the AFIO registers,  but this doesn't work if the clock for the gpio bank hasn't been enabled

Reply:
+1 I just noticed the same. s/POST_KERNEL/PRE_KERNEL_1/

Reply:
Fixed now via `s/POST_KERNEL/PRE_KERNEL_1/` - thanks!